discovery_engine_core/core/src/engine.rs

```diff
@@ -92,14 +113,63 @@ impl Engine {
             .select(self.stacks.values_mut().collect(), max_documents)
             .map_err(|e| e.into())
     }
+
+    /// Process the feedback about the user spending some time on a document.
+    pub fn time_logged(&mut self, time_logged: &TimeLogged) -> Result<(), Error> {
+        self.ranker.time_logged(time_logged)?;
+
+        ranker_updated(self.stacks.values_mut(), &self.ranker)
+    }
+
+    /// Process the feedback about the user reacting to a document.
+    pub fn user_reacted(&mut self, reaction: &UserReacted) -> Result<(), Error> {
+        let stack = self
+            .stacks
+            .get_mut(&reaction.stack_id)
+            .ok_or(Error::InvalidStackId(reaction.stack_id))?;
+
+        stack.update_relevance(reaction.reaction);
+
+        self.ranker.user_reacted(reaction)?;
+
+        ranker_updated(self.stacks.values_mut(), &self.ranker)
+    }
+}
+
+/// The ranker has been updated.
+///
+/// The ranker could rank the documents in a different order so we update the stacks with it.
+fn ranker_updated<'a, R: Ranker>(
+    stacks: impl Iterator<Item = &'a mut Stack>,
+    ranker: &R,
+) -> Result<(), Error> {
+    let errors = stacks.fold(vec![], |mut errors, stack| {
+        if let Err(e) = stack.rank(ranker).map_err(Error::StackOpFailed) {
+            errors.push(e);
+        }
+
+        errors
+    });
+
+    if errors.is_empty() {
+        Ok(())
+    } else {
+        Err(Error::Errors(errors))
+    }
 }
 
 /// A wrapper around a dynamic error type, similar to `anyhow::Error`,
 /// but without the need to declare `anyhow` as a dependency.
 pub(crate) type GenericError = Box<dyn std::error::Error + Sync + Send + 'static>;
 
 /// Provides a method for ranking slice of [`Document`] items.
-pub(crate) trait Ranker {
+pub trait Ranker {
     /// Performs the ranking of [`Document`] items.
     fn rank(&self, items: &mut [Document]) -> Result<(), GenericError>;
+
+    /// Learn from the time a user spent on a document.
```

Comment:
general question: here "Learn" as in "machine learn", right?

Reply:
yes